Catalog Filter feature

troycarpenter opened the pull request:
Backend and GUI code for the new catalog filter feature.  This feature supersedes the private catalog feature.  Instead of restricting a catalog to a single user, this feature can be used to customize which catalogs a user will be able to use.

Upon upgrade, the system will create a DEFAULT filter (or access) group which includes all current catalogs.  Each user will be assigned to the DEFAULT group.  On a fresh install, the DEFAULT group will be empty.  Newly created catalogs are not automatically added to any catalog filters and must be added manually after creation. 

Admins can then create any number of filter groups with any number of included catalogs, then assign one of the groups to each user.  Any catalog created after that will need to be added to a filter group for other to access.

This feature is turned on with the config setting "catalog_filter = true".  If false, everyone will have access to all catalogs.

Closes #2993 
Closes #3290

troycarpenter commented:
Wow.  Looking at the Code Climate report.  A lot of that code is complex for sure (and a few routines I made more complex).  It looks like it dings even the older stuff that was already there.  I do notice that were some tabs instead of spaces, I thought I got all those before submitting.  Probably some other things from my code as well.

lachlan-00 commented:
don't forget the php codestyle with the scripts in `/resources/scripts/tests/`

i usually just run `sh ./resources/scripts/tests/*.sh`

you can also manually run it `php ./vendor/bin/php-cs-fixer fix -vvvv`

troycarpenter commented:
I was not aware of those scripts and will have a look.  This submission is more so you can see what's happening on the backend and maybe have a good idea how to represent it on the front end.

troycarpenter commented:
Just pushed to code to allow selection of the catalog filter on user creation, as well as allowing editing of the filter on a user.  Verified both by checking that the database value for catalog_access_group in the user table was correct after user creation and changed when the user was edited.

lachlan-00 commented:
there'll be some conflict from the user_filter additional checks i added but i'll look at actually using the branch myself and working on the UI/fixes needed to get it in

troycarpenter commented:
> there'll be some conflict from the user_filter additional checks i added but i'll look at actually using the branch myself and working on the UI/fixes needed to get it in

I just added a new section in the discussion thread about the architecture and how the new tables need to be filled in when creating a new filter or catalog.

lachlan-00 commented:
ok, that's good now, i'm going to finish for the day here with one last major merge.

I'll look at copying your branch to my test server and playing there

it doesn't seem to affect your files but i'm starting the subsonic data rewrite. The first part is here 45c3e3c41

troycarpenter commented:
I was pleasantly surprised when I didn't really have to do anything with Subsonic API to have all this work, so I'm expecting any rewrite will also have minimal effects.

troycarpenter commented:
Changed the status back to WIP because I found a problem assigning the catalog to the user in the edit user screen.

lachlan-00 commented on resources/sql/ampache.sql:
```diff
@@ -264,6 +264,53 @@ CREATE TABLE IF NOT EXISTS `catalog` (
 
 -- --------------------------------------------------------
 
+--
```
drop the sql file from this as it's not relevant til i make a release and i do it from a zip install

troycarpenter replied:
Just to be clear, I needed this file to test a clean install, but not submitted in the pull request?

lachlan-00 replied:
yeah, the sql file only gets updated for new people and it's sort of a new install only thing i need to worry about.

the update.php file is more important because i generate the sql file from a new install to create a default

troycarpenter replied:
Ok,  I will make that change and some others when I have time today.

lachlan-00 commented on src/Repository/SongRepository.php:
```diff
@@ -91,9 +105,15 @@ public function getRandomByArtist(
         $sql .= "ORDER BY RAND()";
 
         $db_results = Dba::read($sql, array($artist->getId()));
+//        debug_event(self::class, 'getRandomByArtist ' . $sql, 5);
```
make sure the pull is set to allow edits cause i'll basically hit it and reupload things that are out of place

troycarpenter replied:
The checkbox on the right side is checked.

Those commented out debug_events appear in a lot of places,. btw.

lachlan-00 commented:
one of the reasons the initial catalog_map table stuff took me so long was how slow some things were but it really just requires a bit of usage and you pick them up as you go.

I'm now finally healthy from family kid illnesses to actually look at this again, i've pulled and merged to a local develop branch for me to work on.

troycarpenter commented:
I'm glad you and your family are well again.

Let me know how the testing goes for you.  There are some improvements I would like to make once the functionality is there.  For instance, adding the ability to add the catalog to the appropriate access filters directly on the add/edit catalog screens to avoid having to go to the edit catalog filter screen if necessary.

And I think the filter database routines can be pulled out of the catalog class and put in their own...not sure if there needs to be a completely new class or not, though.

And the whole mess can be cleaned up...like I said before...I discovered right at the end how to be more efficient with the database calls.

In any case, the learning took a bit more time from work and family than I expected and I need to pay more attention there.  But on the bright side, I'm much more familiar with the code and should be better able to contribute going forward.

lena-wei commented:
I hope this is the right place to comment this. 
I have a fresh install of ampache 5.4.0. I set catalog_filter to "true". 
Now when I add a catalog, the input field for catalog user does not show up.
![Screenshot_20220524 16_0854-Ampache __ For the Love of Music - Home — Mozilla Firefox](https://user-images.githubusercontent.com/106166128/170056034-8d821377-9245-4274-90cc-26c47b662154.png)

troycarpenter commented:
> I hope this is the right place to comment this. I have a fresh install of ampache 5.4.0. I set catalog_filter to "true". I did nothing else. Now when I add a catalog, the input field for catalog user does not show up.
> 
> Any idea how to fix this?
> 
> 

You are best to start a new issue.  The feature addressed by this pull request is not in the code base yet, so it would not affect what you posted.  I have looked at my non-modified version and that field is also not working, so it is a bug.

However, be aware that this will totally replace the functionality you are trying to use, and greatly expands its capabilities.  It may not be worth it to the developers to fix that given this feature will shortly make fixing it moot.  

If you want to read about how the new feature works and its development trail, you can read the discussion while I developed it here: https://github.com/ampache/ampache/discussions/3244

I'm waiting for this expanded feature to be added to the codebase before I start working on improving it.

lena-wei commented:
@troycarpenter  Thank you for that instant reply, I see now that I'm disturbing your pull request thread, sorry for the mess! I won't make a new issue for the obsolete catalog_filter then. 

I've read your discussion and I'm very excited about the new feature. Thank you for the great work! 

I'd love to help with testing on user level. (My kids will too and they are the most random input source ;) ). Can you let me know when it's a good moment to do so and on which branch?

troycarpenter commented:
> @troycarpenter Thank you for that instant reply, I see now that I'm disturbing your pull request thread, sorry for the mess! I won't make a new issue for the obsolete catalog_filter then.
> 
> I've read your discussion and I'm very excited about the new feature. Thank you for the great work!
> 
> I'd love to help with testing on user level. (My kids will too and they are the most random input source ;) ). Can you let me know when it's a good moment to do so and on which branch?

I believe the new code is in the process of being better integrated into the base code.  Given that it deals with databases, it would be better to wait until the code is merged and there's a new release including it.  That way if something changes during the merge, you won't have a potential database inconsistency.

I haven't moved my production server to my own code yet for the same reason.  I'm also waiting to move all my users to Ampache until this is merged because I can't control access to certain catalogs (for my kids) without it.

lena-wei commented:
> I believe the new code is in the process of being better integrated into the base code. Given that it deals with databases, it would be better to wait until the code is merged and there's a new release including it. That way if something changes during the merge, you won't have a potential database inconsistency. 
> 
> I haven't moved my production server to my own code yet for the same reason. I'm also waiting to move all my users to Ampache until this is merged because I can't control access to certain catalogs (for my kids) without it.

I see. I'll try to be patient then. Have a good day!

lachlan-00 commented:
hey, i'm again, going to start on this, i've had a lot of work to get done professionally and will be back onto this one now

lachlan-00 commented on src/Module/System/Update.php:
```diff
@@ -4379,4 +4382,112 @@ public static function update_540002(): bool
 
         return $retval;
     }
+
```
rename these to 55xxx and it's duplicated?

lachlan-00 replied:
550001 would be fine

lachlan-00 commented on src/Module/System/Update.php.orig:
```diff
@@ -0,0 +1,4306 @@
+<?php
```
can delete this one

lachlan-00 commented on src/Module/System/Update.php:
```diff
@@ -4379,4 +4382,112 @@ public static function update_540002(): bool
+        $retval &= (Dba::write($sql) !== false);
+
+        // Add the new catalog_access table
+        $sql = "CREATE TABLE IF NOT EXISTS `catalog_access` (`access_group_id` int(11) UNSIGNED NOT NULL,`catalog_id` int(11) UNSIGNED NOT NULL,`enabled` tinyint(1) UNSIGNED NOT NULL DEFAULT 0,FOREIGN KEY (`access_group_id`) REFERENCES catalog_access_group(`id`) ON DELETE CASCADE, FOREIGN KEY (`catalog_id`) REFERENCES catalog(`id`) ON DELETE CASCADE) ENGINE=$engine DEFAULT CHARSET=$charset COLLAT ...
```
errors making that table. i deleted the second update so maybe that's incorrect compared to the first db update?

```
Error_query SQL: CREATE TABLE IF NOT EXISTS `catalog_access` (`access_group_id` int(11) UNSIGNED NOT NULL, `catalog_id` int(11) UNSIGNED NOT NULL, `enabled` tinyint(1) UNSIGNED NOT NULL DEFAULT 0,FOREIGN KEY (`access_group_id`) REFERENCES catalog_access_group(`id`) ON DELETE CASCADE, FOREIGN KEY (`catalog_id`) REFERENCES catalog(`id`) ON DELETE CASCADE) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```

```
Error_query MSG: ["HY000",1005,"Can't create table `ampachetest`.`catalog_access` (errno: 150 \"Foreign key constraint is incorrectly formed\")"]
```

troycarpenter replied:
I'm not seeing this problem.  I copied the code and used it directly in mysql:
```
MariaDB [(none)]> use ampache
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [ampache]> CREATE TABLE IF NOT EXISTS `catalog_access_group` (`id` int(11) UNSIGNED NOT NULL AUTO_INCREMENT,`name` varchar(128) COLLATE utf8mb4_unicode_ci DEFAULT NULL,PRIMARY KEY (`id`),UNIQUE KEY `name` (`name`)) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
Query OK, 0 rows affected (0.029 sec)

MariaDB [ampache]> CREATE TABLE `default_catalog_access_group` (`default_catalog_access_group_id` int(11) UNSIGNED PRIMARY KEY,FOREIGN KEY (default_catalog_access_group_id) references catalog_access_group (id));
Query OK, 0 rows affected (0.046 sec)

MariaDB [ampache]> CREATE TABLE IF NOT EXISTS `catalog_access` (`access_group_id` int(11) UNSIGNED NOT NULL, `catalog_id` int(11) UNSIGNED NOT NULL, `enabled` tinyint(1) UNSIGNED NOT NULL DEFAULT 0,FOREIGN KEY (`access_group_id`) REFERENCES catalog_access_group(`id`) ON DELETE CASCADE, FOREIGN KEY (`catalog_id`) REFERENCES catalog(`id`) ON DELETE CASCADE) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
Query OK, 0 rows affected (0.030 sec)
```
The tables have to be created in that order or else the KEY references don't work.  I'm going to restore the database and try it from the code,.

lachlan-00 replied:
ok, that's good, it could also be from me and my database. i'll do some more tesitng tomorrow but at least we're on our way!

lachlan-00 commented:
not sure what i'm getting with the foreign keys but they're still erroring.

is the foreign key being used for cleanup? of removed catalogs and groups?

troycarpenter commented:
Yes, they are used for cleaning up the database when either a filter or a catalog is deleted.  If the filter is deleted, then all database entries for that filter are deleted.  If a catalog is deleted, then all filter entries for that catalog get deleted.

lachlan-00 commented:
are you using mariadb or mysql? might be differences if you're on mysql

troycarpenter commented:
> are you using mariadb or mysql? might be differences if you're on mysql
```
www-data@ampache-testing:~/ampache/src/Module/System$ mysql --version
mysql  Ver 15.1 Distrib 10.6.7-MariaDB, for debian-linux-gnu (x86_64) using  EditLine wrapper
```

lachlan-00 commented:
i'm using 10,5 mariadb. i wonder if there are indexes or contstraints i've got on my DB that are making it fail.

it might be easier to use garbage collection instead of the cascade

troycarpenter commented:
I'm not sure how garbage collection works, and I'm not certain of the effect of leaving entries in the database that are no longer valid.  These tables are used pretty extensively to select and exclude items both in the GUI and in the background.

The problem I have is that I can't reproduce the issue so it makes it hard for me to troubleshoot to find out how to solve it.

troycarpenter commented:
So what do we need to do to move forward on this?  I don't know why the key restrictions are causing an issue.  Let me do a clean install, which should install an older database then do an upgrade and see if the issue shows up there.

lachlan-00 commented:
very weird, i'm going to take the key checks out for me and just work on it from there

lachlan-00 commented:
as long as i can use it without the foreign key checks i can fill in the parts for removal/garbage collection

troycarpenter commented:
I'm not against you making changes to get it to work for you, especially if that makes it work universally for everyone.  I certainly don't know how to fix something I don't see.

lachlan-00 commented:
ok, again, i've been away for work a lot (the perks of having a high pressure job that's not always pressure)

but i'll be trying to get back to fulltime ampache next week
i'm going to fork your branch into my personal account and run it up properly.

troycarpenter commented:
Do you have an update on the progress on this pull request?

lachlan-00 commented:
it's my next todo, with work i've been left with no time to do anything on the side lately but i've forked and this week i'll have more time

lachlan-00 commented:
going to move this into a sub branch. going to probably make a lot of changes but this will keep all your commits in there as is.

lachlan-00 commented:
@troycarpenter might keep the conversation here while i work in the branch.

i see `default_catalog_access_group` but i don't really see a purpose? i've put the default group back from 1 to 0 as it's easier to cast to an int and get 0 instead of relying on a query

troycarpenter commented:
The purpose of that is to prevent someone from deleting all the groups.  At least one group must exist for everything to work correctly.  The database checks prevent the default group from being deleted when it's still referenced in that table.  It's a safeguard.

Sent from Nine

________________________________
From: lachlan ***@***.***>
Sent: Thursday, July 21, 2022 12:08 AM
To: ampache/ampache
Cc: Troy Carpenter; Mention
Subject: Re: [ampache/ampache] Catalog Filter feature (PR #3260)

@troycarpenter might keep the conversation here while i work in the branch.
i see default_catalog_access_group but i don't really see a purpose? i've put the default group back from 1 to 0 as it's easier to cast to an int and get 0 instead of relying on a query
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

lachlan-00 commented:
hey, just want to say this have been really nice to work on. you put a lot of effort in here and it's appreciated

lachlan-00 commented:
I'm also going to be a pain and rename these tables before develop (sorry if your main server is running this)

using the prefix catalog_filter  we have group:
![image](https://user-images.githubusercontent.com/1305249/180700308-6479f9ac-b1e5-492c-942e-c20131ebee17.png)

and group_map:
![image](https://user-images.githubusercontent.com/1305249/180700841-d0385a3f-a900-4bbb-864c-1c9870e4dda8.png)

This also renames the column in the user table to catalog_filter_group to match

As before the default group is now 0. which allows dropping that default table.

so far though i'm pretty happy with this and was able to group myself and a user and effectively separate them.

Going on my main server now!

lachlan-00 commented:
moving to a pull now. just have to test and try to break before develop
https://github.com/ampache/ampache/pull/3300
